File: apps/frontend/src/lib/api/audit.ts
import { client } from './client';
import type { Employee } from './auth';

export interface AuditCycle {
  id: string;
  name: string;
  scope_department_id: string | null;
  scope_location: string | null;
  start_date: string;
  end_date: string;
  status: 'Draft' | 'Active' | 'Closed';
  created_by: string;
  closed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface AuditCycleDetail extends AuditCycle {
  scope_department_name: string | null;
  created_by_name: string | null;
  assigned_auditors: Employee[];
  item_count: number;
  verified_count: number;
  missing_count: number;
  damaged_count: number;
}

export interface AuditItem {
  id: string;
  audit_cycle_id: string;
  asset_id: string;
  auditor_id: string | null;
  result: 'Verified' | 'Missing' | 'Damaged' | null;
  notes: string | null;
  verified_at: string | null;
  created_at: string;
  updated_at: string;
  asset_tag: string;
  asset_name: string;
  asset_status: string;
  asset_location: string | null;
}

export interface DiscrepancyReport {
  id: string;
  audit_cycle_id: string;
  asset_id: string;
  audit_item_id: string;
  issue_type: 'Missing' | 'Damaged';
  resolved: boolean;
  resolved_by: string | null;
  resolved_at: string | null;
  created_at: string;
  updated_at: string;
  cycle_name: string;
  asset_tag: string;
  asset_name: string;
  resolved_by_name: string | null;
}

export interface CreateCycleRequest {
  name: string;
  scope_department_id?: string | null;
  scope_location?: string | null;
  start_date: string;
  end_date: string;
}

export interface CreateCycleResponse {
  audit_cycle: AuditCycle;
}

export interface ListCyclesResponse {
  audit_cycles: AuditCycleDetail[];
}

export interface GetCycleResponse {
  audit_cycle: AuditCycleDetail;
}

export interface ListItemsResponse {
  items: AuditItem[];
}

export interface AssignAuditorsRequest {
  employee_ids: string[];
}

export interface PatchItemRequest {
  result?: 'Verified' | 'Missing' | 'Damaged' | null;
  notes?: string | null;
}

export interface ListDiscrepancyReportsParams {
  cycle_id?: string;
  resolved?: string;
}

export interface ListDiscrepancyReportsResponse {
  discrepancy_reports: DiscrepancyReport[];
}

export const createCycle = async (req: CreateCycleRequest): Promise<CreateCycleResponse> => {
  const { data } = await client.post('/v1/audit-cycles', req);
  return data;
};

export const listCycles = async (): Promise<ListCyclesResponse> => {
  const { data } = await client.get('/v1/audit-cycles');
  return data;
};

export const getCycle = async (id: string): Promise<GetCycleResponse> => {
  const { data } = await client.get(`/v1/audit-cycles/${id}`);
  return data;
};

export const assignAuditors = async (cycleId: string, req: AssignAuditorsRequest): Promise<{ message: string }> => {
  const { data } = await client.post(`/v1/audit-cycles/${cycleId}/auditors`, req);
  return data;
};

export const listItems = async (cycleId: string, myItems?: boolean): Promise<ListItemsResponse> => {
  const params: Record<string, string> = {};
  if (myItems) params.my_items = 'true';
  const { data } = await client.get(`/v1/audit-cycles/${cycleId}/items`, { params });
  return data;
};

export const patchItem = async (itemId: string, req: PatchItemRequest): Promise<{ message: string }> => {
  const { data } = await client.patch(`/v1/audit-items/${itemId}`, req);
  return data;
};

export const closeCycle = async (cycleId: string): Promise<{ message: string }> => {
  const { data } = await client.patch(`/v1/audit-cycles/${cycleId}/close`);
  return data;
};

export const listDiscrepancyReports = async (params?: ListDiscrepancyReportsParams): Promise<ListDiscrepancyReportsResponse> => {
  const { data } = await client.get('/v1/discrepancy-reports', { params });
  return data;
};

export const resolveDiscrepancy = async (id: string): Promise<{ message: string }> => {
  const { data } = await client.patch(`/v1/discrepancy-reports/${id}/resolve`);
  return data;
};

File: apps/backend/internals/models/audit.go
package models

import "time"

type AuditCycle struct {
	ID                 string     `json:"id"`
	Name               string     `json:"name"`
	ScopeDepartmentID  *string    `json:"scope_department_id"`
	ScopeLocation      *string    `json:"scope_location"`
	StartDate          string     `json:"start_date"`
	EndDate            string     `json:"end_date"`
	Status             string     `json:"status"`
	CreatedBy          string     `json:"created_by"`
	ClosedAt           *time.Time `json:"closed_at"`
	CreatedAt          time.Time  `json:"created_at"`
	UpdatedAt          time.Time  `json:"updated_at"`
}

type AuditCycleDetail struct {
	AuditCycle
	ScopeDepartmentName *string `json:"scope_department_name"`
	CreatedByName       *string `json:"created_by_name"`
	AssignedAuditors    []Employee `json:"assigned_auditors,omitempty"`
	ItemCount           *int    `json:"item_count"`
	VerifiedCount       *int    `json:"verified_count"`
	MissingCount        *int    `json:"missing_count"`
	DamagedCount        *int    `json:"damaged_count"`
}

type AuditItem struct {
	ID           string     `json:"id"`
	AuditCycleID string     `json:"audit_cycle_id"`
	AssetID      string     `json:"asset_id"`
	AuditorID    *string    `json:"auditor_id"`
	Result       *string    `json:"result"`
	Notes        *string    `json:"notes"`
	VerifiedAt   *time.Time `json:"verified_at"`
	CreatedAt    time.Time  `json:"created_at"`
	UpdatedAt    time.Time  `json:"updated_at"`
}

type AuditItemDetail struct {
	AuditItem
	AssetTag     string  `json:"asset_tag"`
	AssetName    string  `json:"asset_name"`
	AssetStatus  string  `json:"asset_status"`
	AssetLocation *string `json:"asset_location"`
}

type DiscrepancyReport struct {
	ID            string     `json:"id"`
	AuditCycleID  string     `json:"audit_cycle_id"`
	AssetID       string     `json:"asset_id"`
	AuditItemID   string     `json:"audit_item_id"`
	IssueType     string     `json:"issue_type"`
	Resolved      bool       `json:"resolved"`
	ResolvedBy    *string    `json:"resolved_by"`
	ResolvedAt    *time.Time `json:"resolved_at"`
	CreatedAt     time.Time  `json:"created_at"`
	UpdatedAt     time.Time  `json:"updated_at"`
}

type DiscrepancyReportDetail struct {
	DiscrepancyReport
	CycleName  string  `json:"cycle_name"`
	AssetTag   string  `json:"asset_tag"`
	AssetName  string  `json:"asset_name"`
	ResolvedByName *string `json:"resolved_by_name"`
}
